bouncerscript/test/test_utils.py

```diff
@@ -14,3 +36,198 @@ def test_load_json_from_file():
             json.dump(json_object, f)
 
         assert load_json(output_file) == json_object
+
+
+# api_call {{{1
+@pytest.mark.parametrize("retry_config", ((
+    None
+), (
+    {'retry_exceptions': aiohttp.ClientError}
+)))
+@pytest.mark.asyncio
+async def test_api_call(context, mocker, retry_config):
+    mocker.patch.object(butils, '_do_api_call', new=noop_async)
+    mocker.patch.object(sutils, 'retry_async', new=noop_async)
+    await api_call(context, 'dummy-route', {}, retry_config)
+
+
+# _do_api_call {{{1
+@pytest.mark.parametrize("data,credentials", ((
+    {}, False
+), (
+    {'product': 'dummy'}, True
+)))
+def test_do_successful_api_call(context, mocker, event_loop, fake_session, data, credentials):
+    context.server = get_task_server(context.task, context.config)
+    context.action = get_task_action(context.task, context.config)
+    context.session = fake_session
+
+    if not credentials:
+        del context.config["bouncer_config"][context.server]["username"]
+        del context.config["bouncer_config"][context.server]["password"]
+
+        with pytest.raises(KeyError):
+            response = event_loop.run_until_complete(
+                _do_api_call(context, 'dummy', data)
+            )
+
+        return
+
+    response = event_loop.run_until_complete(
+        _do_api_call(context, 'dummy', data)
+    )
+
+    assert response == '{}'
+    context.session.close()
```

Comment:
Question for my own curiosity: Would it be possible to let the fixture (https://github.com/mozilla-releng/bouncerscript/pull/12/files#diff-3d6daaf40d168d2ef7c9b7e2fa652ff7R48) close the context? I don't know if that's feasible

Reply:
I don't think you could since not all the `context` objects have a session defined.